### PR TITLE
Revert stake unit tests

### LINKED ISSUE #425: Gateway: Unit test for EIP20Gateway revertStake()

### PR DESCRIPTION
**Pull request by sarvesh-ost**
Fixes #425  

Unit test cases for revert stake and minor refactoring.

### CODE REVIEW

File: test/gateway/eip20_gateway/revert_stake.js [hunk 1]
```diff
@@ -52,9 +66,254 @@ contract('EIP20Gateway.revertStake()', function (accounts) {
...
+      {from: stakeMessage.staker},
+    );
+
+    let gatewayFinalBalance = await baseToken.balanceOf(gateway.address);
+    let stakerFinalBalance = await baseToken.balanceOf(stakeMessage.staker);
+
+    assert.strictEqual(
+      gatewayFinalBalance.eq(gatewayInitialBalance.add(penalty)),
+      true,
+      'Panelty must be transferred to gateway',
```
Comment (gulshanvasnani) [inline test/gateway/eip20_gateway/revert_stake.js]: Panelty => 'Penalty'

File: test/gateway/eip20_gateway/revert_stake.js [hunk 2]
```diff
@@ -52,9 +66,254 @@ contract('EIP20Gateway.revertStake()', function (accounts) {
...
+    );
+
+    assert.strictEqual(
+      stakerFinalBalance.eq(stakerInitialBalance.sub(penalty)),
+      true,
+      'Panelty must be transferred from staker',
```
Comment (gulshanvasnani) [inline test/gateway/eip20_gateway/revert_stake.js]: Same as above

File: test/gateway/eip20_gateway/revert_stake.js (hunk 1 above)
Comment (gulshanvasnani) [inline test/gateway/eip20_gateway/revert_stake.js]: Panelty => 'Penalty'

File: test/gateway/eip20_gateway/revert_stake.js (hunk 2 above)
Comment (gulshanvasnani) [inline test/gateway/eip20_gateway/revert_stake.js]: Same as above

File: contracts/gateway/GatewayBase.sol [hunk 3]
```diff
@@ -521,6 +521,22 @@ contract GatewayBase is Organized {
...
+     *
+     * @return penalty_ Amount of penalty needs to be paid by message initiator
+     *                  to revert message transfers.
+     */
+    function penalty(uint256 _bounty)
+        internal
+        pure
+        returns(uint256 penalty_)
+    {
+        penalty_ = _bounty.mul(REVOCATION_PENALTY).div(100);
```
Comment (deepesh-kn) [inline contracts/gateway/GatewayBase.sol]: Find the usage of `REVOCATION_PENALTY`.
This code needs to be update in `progressRevertRedeem` and `progressRevertStake`.

File: contracts/gateway/EIP20Gateway.sol [hunk 4]
```diff
@@ -505,17 +505,14 @@ contract EIP20Gateway is GatewayBase {
...
+            "Message hash must not be zero."
         );
 
-        // get the message object for the _messageHash
         MessageBus.Message storage message = messages[_messageHash];
 
-        require(message.sender == msg.sender, "Only staker can revert stake.");
-
         require(
-            message.intentHash != bytes32(0),
```
Comment (deepesh-kn) [inline contracts/gateway/EIP20Gateway.sol]: Why is this deleted?

File: test/gateway/eip20_gateway/revert_stake.js [hunk 5]
```diff
@@ -52,9 +66,254 @@ contract('EIP20Gateway.revertStake()', function (accounts) {
...
+    );
+
+  });
+
+  it('should emit event on success', async function () {
+
+    let penalty = new BN(150);
+
+    await baseToken.approve(gateway.address, penalty, {from: stakeMessage.staker});
+    await gateway.setOutboxStatus(
```
Comment (deepesh-kn) [inline test/gateway/eip20_gateway/revert_stake.js]: new line.

File: test/gateway/eip20_gateway/revert_stake.js [hunk 6]
```diff
@@ -52,9 +66,254 @@ contract('EIP20Gateway.revertStake()', function (accounts) {
...
+      stakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    let tx = await gateway.revertStake(
```
Comment (deepesh-kn) [inline test/gateway/eip20_gateway/revert_stake.js]: Add asserts for `gateway.revertStake.call` returned values.
or create a new `it` for this .

File: test/gateway/eip20_gateway/revert_stake.js [hunk 7]
```diff
@@ -52,9 +66,254 @@ contract('EIP20Gateway.revertStake()', function (accounts) {
...
+    );
   });
 
-  it('should fail when messagehash is zero', async function () {
+  it('should fail for zero message hash', async function () {
+
+    let penalty = new BN(150);
+
+    await baseToken.approve(gateway.address, penalty, {from: stakeMessage.staker});
+    await gateway.setOutboxStatus(
```
Comment (deepesh-kn) [inline test/gateway/eip20_gateway/revert_stake.js]: new line.

File: test/gateway/eip20_gateway/revert_stake.js [hunk 8]
```diff
@@ -52,9 +66,254 @@ contract('EIP20Gateway.revertStake()', function (accounts) {
...
+      'Message hash must not be zero.'
+      );
+
+  });
+
+  it('should fail if revocation is already declared', async function () {
+
+    let penalty = new BN(150);
+
+    await baseToken.approve(gateway.address, penalty, {from: stakeMessage.staker});
```
Comment (deepesh-kn) [inline test/gateway/eip20_gateway/revert_stake.js]: Following code is used in all test cases, this can be moved to a function.
```
    await baseToken.approve(gateway.address, penalty, {from: stakeMessage.staker});
    await gateway.setOutboxStatus(
      stakeMessage.messageHash,
      MessageStatusEnum.Declared,
    );
```

File: test/gateway/eip20_gateway/revert_stake.js (hunk 5 above)
Comment (sarvesh-ost) [reply test/gateway/eip20_gateway/revert_stake.js]: why?

**Comment by sarvesh-ost** [comment]
> `should fail if staker balance is less than penalty amount.`

There is already one test `should fail if gateway is not approved for penalty amount`. For `revertStake`, I wouldn't care about staker base taken balance. If staker doesn't approve enough or doesn't have enough balance is similar case for us(As it just fails transferFrom).

File: contracts/gateway/EIP20Gateway.sol (hunk 4 above)
Comment (sarvesh-ost) [reply contracts/gateway/EIP20Gateway.sol]: Because that was duplicate, this check `        require(message.sender == msg.sender, "Only staker can revert stake.");	` is sufficient to validate if message exists.

Comment (schemar) [reply contracts/gateway/EIP20Gateway.sol]: In that case do you require the following?
```solidity
        require(
            _messageHash != bytes32(0),
            "Message hash must not be zero"
        );
```

File: contracts/gateway/GatewayBase.sol [hunk 9]
```diff
@@ -521,6 +521,22 @@ contract GatewayBase is Organized {
...
+    /**
+     * @notice Calculates the penalty amount for reverting a message transfer.
+     *
+     * @param _bounty The amount that facilitator has staked to initiate the
+     *                message transfers.
+     *
+     * @return penalty_ Amount of penalty needs to be paid by message initiator
+     *                  to revert message transfers.
+     */
+    function penalty(uint256 _bounty)
```
Comment (schemar) [inline contracts/gateway/GatewayBase.sol]: I don't think `penalty` is a very good function name here. It lacks context. The function returns the penalty based on the bounty.
Maybe instead name it something like `penaltyFromBounty` or `bountyToPenalty` or `calculatePenalty` or something similar?

File: test/gateway/eip20_gateway/revert_stake.js [hunk 10]
```diff
@@ -52,9 +66,289 @@ contract('EIP20Gateway.revertStake()', function (accounts) {
       burner,
     );
 
+    let hashLockObj = Utils.generateHashLock();
+
+    stakeMessage.hashLock = hashLockObj.l;
+    stakeMessage.unlockSecret = hashLockObj.s;
```
Comment (schemar) [inline test/gateway/eip20_gateway/revert_stake.js]: `l` and `s` are horrible field names 😱

File: test/gateway/eip20_gateway/revert_stake.js [hunk 11]
```diff
@@ -52,9 +66,289 @@ contract('EIP20Gateway.revertStake()', function (accounts) {
...
+      MessageStatusEnum.Declared,
+    );
+
+    await Utils.expectRevert(
+      gateway.revertStake(
+        Utils.ZERO_BYTES32,
+        {from: stakeMessage.staker},
+      ),
+      'Message hash must not be zero.'
+      );
```
Comment (schemar) [inline test/gateway/eip20_gateway/revert_stake.js]:
```suggestion
    );
```

File: test/gateway/eip20_gateway/revert_stake.js [hunk 12]
```diff
@@ -52,9 +66,289 @@ contract('EIP20Gateway.revertStake()', function (accounts) {
...
+      stakeMessage.messageHash,
+      MessageStatusEnum.Declared,
+    );
+
+    await Utils.expectRevert(
+      gateway.revertStake(
+        Utils.ZERO_BYTES32,
+        {from: stakeMessage.staker},
+      ),
+      'Message hash must not be zero.'
```
Comment (schemar) [inline test/gateway/eip20_gateway/revert_stake.js]:
```suggestion
      'Message hash must not be zero.',
```

File: test/gateway/eip20_gateway/revert_stake.js (hunk 8 above)
Comment (schemar) [reply test/gateway/eip20_gateway/revert_stake.js]: Maybe if you find a good function name for the function that does these two things, then it is still readable and clear when you make the function call of that function in these places.

File: test/gateway/eip20_gateway/revert_stake.js [hunk 13]
```diff
@@ -52,9 +66,289 @@ contract('EIP20Gateway.revertStake()', function (accounts) {
...
+    await Utils.expectRevert(
+      gateway.revertStake(
+        stakeMessage.messageHash,
+        {from: stakeMessage.staker},
+      ),
+      'Message on source must be Declared.'
+    )
+  });
+
+  it('should fail for undeclared message', async function () {
```
Comment (schemar) [inline test/gateway/eip20_gateway/revert_stake.js]: Is there a test that asserts the behavior in case there is no outbox state set at all for a message? Should there be one?

Comment (sarvesh-ost) [reply test/gateway/eip20_gateway/revert_stake.js]: wouldn't this case be similar to `Undeclared` status, because Message status is an enum with default value as zero of type uint8? 
If outbox doesn't have message status it will return default value which is undeclared (zero).  

What do you think?

File: test/gateway/eip20_gateway/revert_stake.js (hunk 10 above)
Comment (sarvesh-ost) [reply test/gateway/eip20_gateway/revert_stake.js]: Yes, agree!!! This is used across files, my IDE is not smart enough to refactor this easily 😞  

Can we please handle this in a separate ticket?

File: test/gateway/eip20_gateway/revert_stake.js (hunk 11 above)
Comment (sarvesh-ost) [reply test/gateway/eip20_gateway/revert_stake.js]: Done

File: test/gateway/eip20_gateway/revert_stake.js (hunk 12 above)
Comment (sarvesh-ost) [reply test/gateway/eip20_gateway/revert_stake.js]: Done

File: test/gateway/eip20_gateway/revert_stake.js (hunk 8 above)
Comment (sarvesh-ost) [reply test/gateway/eip20_gateway/revert_stake.js]: I couldn't think of function name without `And` in it, do you have suggestion? 

Do you strongly feel this must be fixed?

File: test/gateway/eip20_gateway/revert_stake.js (hunk 10 above)
Comment (schemar) [reply test/gateway/eip20_gateway/revert_stake.js]: Of course. Just wanted to mention it.

File: test/gateway/eip20_gateway/revert_stake.js (hunk 13 above)
Comment (schemar) [reply test/gateway/eip20_gateway/revert_stake.js]: True 👍

File: contracts/gateway/EIP20Gateway.sol (hunk 4 above)
Comment (deepesh-kn) [reply contracts/gateway/EIP20Gateway.sol]: `bytes32(0)` is valid hash but we will not allow `bytes32(0)` hash since it can be due to invalid input data. We can discuss this if we want to allow it.

File: test/gateway/eip20_gateway/revert_stake.js (hunk 5 above)
Comment (deepesh-kn) [reply test/gateway/eip20_gateway/revert_stake.js]: I think we all should follow some styling guide. I add new lines in my code. 
I am fine with not removing it, but we all should follow the same. 👍

File: test/gateway/eip20_gateway/revert_stake.js (hunk 8 above)
Comment (deepesh-kn) [reply test/gateway/eip20_gateway/revert_stake.js]: For writing, test cases don't have a strong view on what should we follow here. I am fine even if it is in this way.

File: test/gateway/eip20_gateway/revert_stake.js (hunk 10 above)
Comment (sarvesh-ost) [reply test/gateway/eip20_gateway/revert_stake.js]: Create ticket 
https://github.com/OpenSTFoundation/mosaic-contracts/issues/597

**Comment by gulshanvasnani** [comment]
LGTM 👍 💯
